proposals/wg-image-compatibility.md

```diff
@@ -0,0 +1,109 @@
+
+Proposal created from [OCI WG template](https://github.com/opencontainers/tob/blob/master/WG-TEMPLATE.md).
+
+## Image Compatibility OCI Working Group - Governance Charter
+
+This document describes the basic governance principles for the Image Compatibility Working Group (the “WG”).
+
+The WG operates as an OCI Working Group under the [Open Container Initiative (OCI) Charter](https://github.com/opencontainers/tob/blob/master/CHARTER.md), which describes the responsibilities of the OCI Technical Oversight Board (the "TOB”).
+The WG is established by the TOB as an OCI Working Group pursuant to the OCI Charter.
+Accordingly, the WG will operate in accordance with the OCI Charter and OCI's other policies and procedures, supplemented by the details below.
+
+## Purpose
+
+In mission-critical industries, applications often require special features provided by host operating systems.
+These are usually described in the release notes or installation guides.
+We believe we have identified the missing puzzle in the OCI standard, the image compatibility, which is a key feature that describes the requirements of special containerized applications that have hard requirements for specific kernel versions, configurations, modules, or out-of-tree drivers.
+
+For example, a container application free5GC UPF requiring the gtp5g kernel module will only be compatible with kernel 5.0.0-23-generic or 5.4.x due to the module's hard kernel version requirements - https://free5gc.org/guide/3-install-free5gc/.
+If SR-IOV or GPU is required by the container, other requirements have to be specified against the host OS.
+Kata Containers use cases are the best example: https://github.com/kata-containers/kata-containers/tree/main/docs/use-cases.
+
+
+The described incompatibility issue applies to all use cases where specific host configuration is required, from bare-metal (e.g. high-performance computing) to distributed systems.
+
+In multi-cloud or hybrid cloud scenarios, one containerized cloud-native application will need to run on different Kubernetes distributions that use different host operating systems on worker nodes.
+
+For bare-metal and cloud operating systems can use different kernel versions with different configurations.
+Different operating system distributions may integrate different devices and their corresponding drivers, even for the same device different versions of drivers may be used.
+Therefore, different operating system distributions are not identical.
+Even within the same OS distro, different releases are also not identical.
+
+An image compatibility would help container image authors describe compatibility requirements in a standard way.
+The specification will be uploaded with the image to the image registry.
+This makes hard container compatibility requirements discoverable, programmable, and will support different consumers and cover use cases where the application requires a specific compatible environment.
```

Comment:
> This appears to follow the idea that exists outside the container world today. Where ISV X certifies their application on a distribution Y, or multiple distributions U, V, and W; runs tests against all those distros and then declares kernel-A, kernel-B, and kernel-C versions are supported. Effectively cutting out all other distributions. Recreating such a "restrictive" environment via a specification may not be in the best interest of the ecosystem as a hole.
> Said differently if I provide an application container and the only means I have to specify requirements are via versions, i.e. kernel-6.2 than there is a very good chance that at least for a period of time I may be cut out of certain platforms until the hosts of those platforms, thinking of EKS, GKE, AKS for example, catch up to my requirement.

I think this is a slightly different use case. It’s just about expressing minimal mandatory requirements for containers in terms of kernel configuration/features that are crucial for containers to run on the host. It’s not about restricting containers to run in the environment in which they were tested, or creating some kind of “certification” chain for different distributions or kernel versions.

> On the contrary if I can specify a feature or interface requirement in some way shape or form then I may still be cut out for some time, but I as the container vendor am not forcing the host environment to do a wholesale upgrade of the kernel, in this example. A wholesale upgrade may be more painful than a feature backport such as a specific network driver.

As a container vendor, if you require specific configuration or kernel modules that are critical to run a containerized application, then as an environment provider you have no choice but to meet such requirements, if you want the container to run there.

> As such the question presents itself if this is not an opportunity to work at the very least with the Linux kernel community of a feature definition system 

The work with the kernel community could be done in parallel, but IMO it's a much more complex topic that would take a lot of time than introducing optional image compatibility schema. We can already express the features over configuration, boot args etc.

> rather than depending on arbitrary version numbers.

Kernel version will not be a mandatory field to set, but may apply to containers requiring modules that have very strict restrictions on kernel version.

I get the impression that you think we want to make compatibility spec to become a hard requirement for images or completely change the way how containers should be launched. We don’t. We simply want to give image authors a chance to describe the critical requirements for containerized applications that they can run on the host. What container consumers do with this information will be up to them.

Reply:
@rjschwei would you like to join the working group?